procinfo/src/macos.rs

```diff
@@ -227,3 +198,121 @@ impl LocalProcessInfo {
         }
     }
 }
+
+fn parse_exe_and_argv_sysctl(buf: Vec<u8>) -> Option<(PathBuf, Vec<String>)> {
+    use libc::c_int;
+
+    // The data in our buffer is laid out like this:
+    // argc - c_int
+    // exe_path - NUL terminated string
+    // argv[0] - NUL terminated string
+    // argv[1] - NUL terminated string
+    // ...
+    // argv[n] - NUL terminated string
+    // envp[0] - NUL terminated string
+    // ...
+
+    let mut ptr = &buf[0..buf.len()];
+
+    let argc: c_int = unsafe { std::ptr::read(ptr.as_ptr() as *const c_int) };
+    ptr = &ptr[std::mem::size_of::<c_int>()..];
+
+    fn consume_cstr(ptr: &mut &[u8]) -> Option<String> {
+        // Trim out trailing nulls
+        let not_nul = ptr.iter().position(|&c| c != 0)?;
+        *ptr = ptr.get(not_nul..)?;
```

Comment:
This comment doesn't match what's happening here: it's skipping leading NULs.  I think this is a bit counterintuitive: I think this function is a bit easier to reason about if we know that `ptr` points to the start of a c-string and then updates ptr to point to the start of the next one when it returns.

I think moving the nul-skipping to the bottom of this function would satisfy that!

Reply:
This required me to add a branch to make sure we haven't run off the end, added!